### DISCUSSION TITLE
Error ```(please help ive been trying to fix this for over 5 months```

### DISCUSSION THREAD

**Discussion by astrogames06** [bevyengine/bevy#5013] [Q&A]
## Bevy version

0.7.0

## OS

WSL2 Ubuntu-18.04

## What you did

i ran ```cargo run```

## Error
```
error: couldn't create a temp dir: Read-only file system (os error 30) at path "/tmp/rustcGvmvDZ"

error: could not compile `wayland-client` due to previous error 

Caused by:
  process didn't exit successfully: `rustc --crate-name wayland_client --edition=2018 /root/.cargo/registry/src/github.com-1ecc6299db9ec823/wayland-client-0.29.4/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="scoped-tls"' --cfg 'feature="use_system_lib"' -C metadata=c9c104fd456eb761 -C extra-filename=-c9c104fd456eb761 --out-dir /mnt/d/rust_bevy_test/target/debug/deps -L dependency=/mnt/d/rust_bevy_test/target/debug/deps --extern bitflags=/mnt/d/rust_bevy_test/target/debug/deps/libbitflags-f1ca410388ec357f.rmeta --extern downcast_rs=/mnt/d/rust_bevy_test/target/debug/deps/libdowncast_rs-728c8f2b8a666e0c.rmeta --extern libc=/mnt/d/rust_bevy_test/target/debug/deps/liblibc-a6a2c08085452da2.rmeta --extern nix=/mnt/d/rust_bevy_test/target/debug/deps/libnix-767d56d797eb760d.rmeta --extern scoped_tls=/mnt/d/rust_bevy_test/target/debug/deps/libscoped_tls-36004976affa6348.rmeta --extern wayland_commons=/mnt/d/rust_bevy_test/target/debug/deps/libwayland_commons-39b12b440c2a190c.rmeta --extern wayland_sys=/mnt/d/rust_bevy_test/target/debug/deps/libwayland_sys-033d0cf33a9890f2.rmeta --cap-lints allow -L native=/usr/lib/x86_64-linux-gnu` (signal: 7, SIGBUS: access to undefined memory)
warning: build failed, waiting for other jobs to finish...      
error: couldn't create a temp dir: Read-only file system (os error 30) at path "/tmp/rustcoBIqVz"

^C  Building [=======================> ] 292/300: wayland-p...  
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo fsck -n -f
sudo: unable to execute /sbin/fsck: Input/output error
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo fsck -n -f     
sudo: unable to execute /sbin/fsck: Input/output error
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo mount -o rw,remount /
mount: /: can't find LABEL=cloudimg-rootfs.
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo mount -o rw,remount /
mount: /: can't find LABEL=cloudimg-rootfs.
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo ntfsfix /dev/sdb1
sudo: unable to execute /bin/ntfsfix: Input/output error
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# sudo mkdir /media/[mount point]
$ sudo mount -o rw /dev/sdb1 /media/[mount point]sudo: unable to execute /bin/mkdir: Input/output error
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# $ sudo mount -o rw /dev/sdb1 /media/[mount point]
-bash: /usr/lib/command-not-found: Input/output error
root@LAPTOP-BOMU0R9Q:/mnt/d/rust_bevy_test# cargo run
   Compiling wayland-protocols v0.29.4
   Compiling wayland-client v0.29.4
   Compiling wayland-cursor v0.29.4
error: couldn't create a temp dir: Read-only file system (os error 30) at path "/tmp/rustcgPKV3u"

error: could not compile `wayland-client` due to previous error 

Caused by:
  process didn't exit successfully: `rustc --crate-name wayland_client --edition=2018 /root/.cargo/registry/src/github.com-1ecc6299db9ec823/wayland-client-0.29.4/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="scoped-tls"' --cfg 'feature="use_system_lib"' -C metadata=c9c104fd456eb761 -C extra-filename=-c9c104fd456eb761 --out-dir /mnt/d/rust_bevy_test/target/debug/deps -L dependency=/mnt/d/rust_bevy_test/target/debug/deps --extern bitflags=/mnt/d/rust_bevy_test/target/debug/deps/libbitflags-f1ca410388ec357f.rmeta --extern downcast_rs=/mnt/d/rust_bevy_test/target/debug/deps/libdowncast_rs-728c8f2b8a666e0c.rmeta --extern libc=/mnt/d/rust_bevy_test/target/debug/deps/liblibc-a6a2c08085452da2.rmeta --extern nix=/mnt/d/rust_bevy_test/target/debug/deps/libnix-767d56d797eb760d.rmeta --extern scoped_tls=/mnt/d/rust_bevy_test/target/debug/deps/libscoped_tls-36004976affa6348.rmeta --extern wayland_commons=/mnt/d/rust_bevy_test/target/debug/deps/libwayland_commons-39b12b440c2a190c.rmeta --extern wayland_sys=/mnt/d/rust_bevy_test/target/debug/deps/libwayland_sys-033d0cf33a9890f2.rmeta --cap-lints allow -L native=/usr/lib/x86_64-linux-gnu` (signal: 7, SIGBUS: access to undefined memory)
warning: build failed, waiting for other jobs to finish...      
error: couldn't create a temp dir: Read-only file system (os error 30) at path "/tmp/rustcn86Z0S"

error: build failed
```

## main.rs

``` rust
use bevy::prelude::*;

fn main() {
    App::new().add_plugins(DefaultPlugins).run();
}
```

## Cargo.toml

``` toml
[package]
name = "rust_bevy_test"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bevy = { version = "0.7.0", features = ["wayland"] }
```

**Comment by bjorn3** [comment]
What does `ls -ld /tmp` show? And `df`? Also given that several system commands give an IO error I suspect that the system running in WSL2 got corrupted somehow. You could try reinstalling it.

**Reply by astrogames06** [reply]
> What does `ls -ld /tmp` show? And `df`? Also given that several system commands give an IO error I suspect that the system running in WSL2 got corrupted somehow. You could try reinstalling it.

i reinstalled wsl2 and when i run `ls -ld /tmp` i get `drwxrwxrwt 2 root root 4096 Jun 15 18:43 /tmp`

**Reply by astrogames06** [reply]
> What does `ls -ld /tmp` show? And `df`? Also given that several system commands give an IO error I suspect that the system running in WSL2 got corrupted somehow. You could try reinstalling it.

## Bevy worked but after changing the main.rs to 
``` rust
use bevy::prelude::*;

fn main() {
    App::new()
        .insert_resource(ClearColor(Color::rgb(0.04, 0.04, 0.04)))
        .insert_resource(WindowDescriptor {
            title: "Bevy".to_string(),
            width: 1280.0,
            height: 720.0,
            ..Default::default()
        })
        .add_plugins(DefaultPlugins)
        .run();
}
```
## i get this error

`thread 'main' panicked at 'Failed to initialize any backend! Wayland status: XdgRuntimeDirNotSet X11 status: XOpenDisplayFailed', /home/jebba/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.1/src/platform_impl/linux/mod.rs:619:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace`

**Comment by alice-i-cecile** [comment]
Can you run other Rust programs successfully?

**Reply by astrogames06** [reply]
> Can you run other Rust programs successfully?

Only programs without crates or programs with very simple crates like Rand.

**Reply by alice-i-cecile** [reply]
Hmm. Can you run Bevy on Windows directly? WSL is sometimes a pain.

**Reply by astrogames06** [reply]
> Hmm. Can you run Bevy on Windows directly? WSL is sometimes a pain.

i dont have rust on windows i found it to hard to install and takes up to much storage

**Comment by hecksadecimal** [comment]
Graphics support for WSL2 isn't quite there yet.
At the moment the vGPU driver only supports OpenGL. 

When either [#841](https://github.com/bevyengine/bevy/issues/841) here or [#7790](https://github.com/microsoft/WSL/issues/7790) on WSL itself gets resolved I suspect there won't be a problem anymore.

**Reply by hecksadecimal** [reply]
Further testing has revealed a workaround, for those determined to develop in WSL.

Steps:
1. Install the mingw toolchain on the WSL instance
2. `rustup target add x86_64-pc-windows-gnu`
3. `cargo build --target x86_64-pc-windows-gnu` or `cargo run --target x86_64-pc-windows-gnu`

If you don't want to specify a target every time, just add this to your `.cargo/config.toml`
```toml
[build]
target = "x86_64-pc-windows-gnu"
```

**Reply by hecksadecimal** [reply]
@bjorn3 @alice-i-cecile
Maybe this quirk and the workaround for it should be documented until it isn't necessary.

**Reply by alice-i-cecile** [reply]
I'm glad to hear that this is working. Can you make a PR to https://github.com/bevyengine/bevy/blob/main/docs/linux_dependencies.md#windows-subsystem-for-linux-wsl-2 ?

**Reply by hecksadecimal** [reply]
Sure can.

**Reply by astrogames06** [reply]
> Further testing has revealed a workaround, for those determined to develop in WSL.
> 
> Steps:
> 
> 1. Install the mingw toolchain on the WSL instance
> 2. `rustup target add x86_64-pc-windows-gnu`
> 3. `cargo build --target x86_64-pc-windows-gnu` or `cargo run --target x86_64-pc-windows-gnu`
> 
> If you don't want to specify a target every time, just add this to your `.cargo/config.toml`
> 
> ```toml
> [build]
> target = "x86_64-pc-windows-gnu"
> ```

its work thank you but where is this .cargo located i cant find it anywhere

**Reply by CatCode79** [reply]
.cargo is an optional configuration folder that you have to create yourself if you don't already have it, more information here: https://doc.rust-lang.org/cargo/reference/config.html
